File: SnapHack/context/samples/Spectacles Mobile Kit/sdk/Android/SpectaclesKit/kit/src/main/java/com/snap/spectacles/kit/SpectaclesAssetRequest.kt
package com.snap.spectacles.kit

import android.net.Uri
import java.util.function.Consumer

sealed interface SpectaclesAssetRequest {

    /**
     * Responsible for providing the Lens with the SpectaclesAsset specified by the uri from the App.
     *
     * @param path The path of the requested SpectaclesAsset.
     * @param version Identify a unique evolution of the asset, if available. This can be either a
     * file checksum or a file timestamp, as determined by the SDK or Lens developer. Pass null to
     * retrieve the most recent asset content.
     * @return The loaded SpectaclesAsset instance.
     */
    data class Load(
        val path: String,
        val version: String?,
        override val onResponse: Consumer<SpectaclesAsset>,
        override val onError: Consumer<SpectaclesRequestException>,
    ) : SpectaclesAssetRequest, SpectaclesRequest.WithResponse<SpectaclesAsset>()
}

File: SnapHack/context/samples/Spectacles Mobile Kit/sdk/iOS/SpectaclesKit/Sources/API/Requests/SpectaclesAssetRequest.swift
// Copyright © 2024 Snap, Inc. All rights reserved.

public import Foundation

public enum SpectaclesAssetRequest: Sendable {
    /// A structure representing an asset with associated data.
    ///
    /// - Parameters:
    ///   - name: The name of the asset.
    ///   - version: Identifies a unique version of the asset, if available. This can be either a
    ///     file checksum or a file timestamp, as determined by the SDK or Lens developer. Pass `nil`
    ///     to retrieve the most recent asset content.
    ///   - data: The raw data associated with the asset.
    public struct Asset: Sendable {
        public var name: String
        public var version: String?
        public var data: Data

        public init(name: String, version: String? = nil, data: Data) {
            self.name = name
            self.version = version
            self.data = data
        }
    }

    public protocol RequestProtocol: SpectaclesRequest.RequestProtocol {}

    case load(any SpectaclesLoadAssetRequest)

    public var underlyingRequest: any RequestProtocol {
        switch self {
        case let .load(request):
            return request
        }
    }
}

public protocol SpectaclesLoadAssetRequest: SpectaclesAssetRequest.RequestProtocol, SpectaclesRequestWithResponseProtocol where Payload == SpectaclesAssetRequest.Asset {
    var uri: String { get }
    var version: String? { get }
}
